packages/server/src/procedure.ts

```diff
@@ -99,12 +101,49 @@ export abstract class Procedure<
     type,
     path,
   }: ProcedureCallOptions<TContext>): Promise<TOutput> {
-    for (const middleware of this.middlewares) {
-      await middleware({ ctx, type, path });
+    const opts = { ctx, type, path };
+
+    const middlewareFns: MiddlewareFunction<TContext>[] = [
+      ...this.middlewares,
+      // wrap the actual resolver and treat as the last "middleware"
+      async (opts) => {
+        try {
+          const input = this.parseInput(rawInput);
+          const data = await this.resolver({ ...opts, input });
+          return {
+            marker: middlewareMarker,
+            ok: true,
+            data,
+          };
+        } catch (_error) {
+          const error = getErrorFromUnknown(_error);
```

Comment:
Without the discriminated union thing, we wouldn't need this try/catch (only to be rethrown later), is that right? Implementation simplicity isn't the _most_ important consideration vs feature set, but maybe a factor.